File: 프로그래머스/2/42586. 기능개발/기능개발.py
def solution(progresses, speeds):
    # 작업 완료까지 걸리는 시간 계산
    done_time = []
    result = []
    
    for i in range(len(progresses)):
        done_num = (100 - progresses[i]) // speeds[i]
        if (100 - progresses[i]) % speeds[i] != 0:
            done_time.append(done_num + 1)
        else:
            done_time.append(done_num)
    
    # 처음부터 큰 수가 나올때까지 pop 
    max_num = done_time[0]
    cnt = 0
    
    for i in range(len(done_time)):
        if done_time[i] <= max_num:
            cnt += 1

        else:
            max_num = done_time[i]
            result.append(cnt)
            cnt = 1
        
        if i == len(done_time) - 1:
            result.append(cnt)
    
    return result

File: 프로그래머스/2/42586. 기능개발/기능개발.java
import java.util.*;

class Solution {
    public int[] solution(int[] progresses, int[] speeds) {
        List<Integer> answer = new ArrayList<>();
        Deque<Integer> q = calcProgressDays(progresses, speeds);

        int maxDay = q.poll();
        int cnt = 1;
        while (!q.isEmpty()) {
            int cur = q.poll();
            
            if (maxDay < cur) {
                maxDay = cur;
                answer.add(cnt);
                cnt = 1;
                continue;
            }
            cnt++;
        }
        
        answer.add(cnt);
        
        return answer.stream().mapToInt(Integer::intValue).toArray();
    }
    
    private Deque<Integer> calcProgressDays(int[] progresses, int[] speeds) {
        int n = progresses.length;
        Deque<Integer> need = new ArrayDeque<>();
        
        for (int i = 0; i < n; i++) {
            int days = (100 - progresses[i]) / speeds[i];
            int left = (100 - progresses[i]) % speeds[i];
            
            if (left == 0) need.offer(days);
            else need.offer(days + 1);
        }
        
        return need;
    }
}
